### ISSUE TITLE
Pylint alerts corrections as part of an intervention experiment

### ISSUE THREAD

**Issue by evidencebp** [python-websockets/websockets#1536]
I'd like to conduct a software engineering experiment regarding the benefit of [Pylint](https://pypi.org/project/pylint/)  alerts removal.
The experiment is described [here](https://github.com/evidencebp/pylint-intervention/).
In the experiments, Pylint is used with some specific alerts, files are selected for intervention and control.
After the interventions are done, one can wait and examine the results.

I'm asking for your approval for conducting an intervention in your repository.

See examples of interventions in [stanford-oval/storm](https://github.com/stanford-oval/storm/pull/181), [gabfl/vault](https://github.com/gabfl/vault/pull/82), and [coreruleset/coreruleset](https://github.com/coreruleset/coreruleset/pull/3837).

You can see the [planed interventions](https://github.com/evidencebp/pylint-intervention/tree/main/interventions/candidates/aaugustin_websockets_interventions_October_05_2024.csv)

@aaugustin , may I do the interventions?

**Comment by aaugustin** [comment]
I understand that the starting point would be https://github.com/evidencebp/pylint-intervention/blob/main/interventions/candidates/aaugustin_websockets_interventions_October_05_2024.csv.

The first line is `too-many-branches`. I'm not interested in refactoring code that works well and that I'm happy with for the sake of making a linter happy. So, the answer is "no".

**Comment by aaugustin** [comment]
I looked at a few other lines and many of them aren't actual problems — e.g. that [wildcard import](https://github.com/python-websockets/websockets/blob/f0d20aafab027e9b99460b193dcb709872b219a5/src/websockets/auth.py#L9) is fine.

**Comment by aaugustin** [comment]
I'll keep this issue open in case I want to run pylint once (blast from 15 years ago!) and see if it catches some interesting stuff.

The hard part isn't running pylint or changing the code; it's making the judgment call of whether the code is OK as it stands or not.

It will be way faster to do it myself directly than by reviewing a pull request.

**Comment by aaugustin** [comment]
What settings are you using? With zero configuration I'm getting:

```
$ pylint src tests | wc -l
    4378
```

which is way more than I'm willing to read (also why this project uses a helpful linter rather than pylint).

**Comment by aaugustin** [comment]
I looked at the first ten lines of your spreadsheet and none of them has value. See the list below for details.

That's a data point for your study in the value of removing pylint warnings :-)

As far as websockets is concerned value is zero.

I recommend that you look at ruff for a fast, modern, pragmatic, Python linter.

---

> src/websockets/asyncio/server.py:111:4: R0912: Too many branches (14/12) (too-many-branches)

code is OK

> src/websockets/server.py:174:15: W0718: Catching too general exception Exception (broad-exception-caught)
> src/websockets/server.py:554:19: W0718: Catching too general exception Exception (broad-exception-caught)

pylint is wrong

> src/websockets/sync/connection.py:317:4: R0915: Too many statements (69/50) (too-many-statements)

code is OK

> src/websockets/client.py:320:19: W0718: Catching too general exception Exception (broad-exception-caught)

pylint is wrong

> src/websockets/sync/server.py:319:0: R0915: Too many statements (65/50) (too-many-statements)

pylint doesn't realize there's a function defined inside this function

> src/websockets/sync/client.py:123:0: R0915: Too many statements (56/50) (too-many-statements)

code is OK

> src/websockets/http11.py:204:4: R0912: Too many branches (15/12) (too-many-branches)

code is acceptable — splitting reading the response body to a separate function and calling that function wouldn't be an clear improvement

> src/websockets/__main__.py:132:11: W0718: Catching too general exception Exception (broad-exception-caught)

pylint is wrong

> src/websockets/legacy/protocol.py:1:0: C0302: Too many lines in module (1641/1000) (too-many-lines)

there's a reason why I rewrote this code... but I'm going to keep the original for 5 years after deprecation — meaning, counting from today — and there's no way I'm going to take the risk of refactoring it

> experiments/broadcast/clients.py:37:11: W0718: Catching too general exception Exception (broad-exception-caught)

this is a one-off experiment; no matter whether the warning is justified, I'm not investing into polishing exception handling

**Comment by evidencebp** [comment]
Thank you very much for your detailed feedback, @aaugustin !

I agree that in your case most alerts are not dramatic.
Part of the experiment is to identify the alerts whos benefit is limited.

As for src/websockets/legacy/protocol.py:1:0: C0302: Too many lines in module (1641/1000) (too-many-lines)
Refactoring it does seems like hell.
Do you think that had you invested in that, it would have been beneficial?

**Comment by aaugustin** [comment]
I decided to refactor because the module was too complex. The module was too long because it was too complex. So there's a relationship between these two facts.

But "it's too long" doesn't give a clue about how to refactor it. I decided to refactor when I learnt about the Sans-I/O design pattern.

It too me 6 years and 2 months from "having the idea to refactor" (cf. #466) to "deprecating the legacy implementation" (yesterday). I'd never embark into 6+ years of work because a module is 1641 lines long. It's nowhere near a large enough problem to warrant 6+ years of work.
